pallets/omnipool/src/lib.rs

```diff
@@ -0,0 +1,1462 @@
+				tradable: Tradability::default(),
+			};
+
+			// Imbalance update and total hub asset liquidity for stable asset first
+			// Note: cannot be merged with native, because the calculations depend on updated values
+			let delta_imbalance = Self::recalculate_imbalance(
+				&((&stable_asset_state, stable_asset_reserve).into()),
+				BalanceUpdate::Decrease(stable_asset_reserve),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+
+			// No imbalance yet, use default value
+			Self::update_imbalance(SimpleImbalance::default(), delta_imbalance)?;
+
+			Self::update_hub_asset_liquidity(&BalanceUpdate::Increase(stable_asset_hub_reserve))?;
+
+			// Imbalance update total hub asset with native asset next
+			Self::recalculate_imbalance(
+				&((&native_asset_state, native_asset_reserve).into()),
+				BalanceUpdate::Decrease(native_asset_reserve),
+			)
+			.ok_or(ArithmeticError::Overflow)?;
+			Self::update_imbalance(<HubAssetImbalance<T>>::get(), delta_imbalance)?;
+
+			Self::update_hub_asset_liquidity(&BalanceUpdate::Increase(native_asset_hub_reserve))?;
+
+			Self::update_tvl(&BalanceUpdate::Increase(
+				native_asset_tvl
+					.checked_add(stable_asset_reserve)
+					.ok_or(ArithmeticError::Overflow)?,
+			))?;
+
+			<Assets<T>>::insert(T::StableCoinAssetId::get(), stable_asset_state);
+			<Assets<T>>::insert(T::NativeAssetId::get(), native_asset_state);
+
+			// Hub asset is not allowed to be bought from the pool
+			<HubAssetTradability<T>>::put(
+				Tradability::SELL | Tradability::ADD_LIQUIDITY | Tradability::REMOVE_LIQUIDITY,
+			);
+
+			Self::deposit_event(Event::TokenAdded {
+				asset_id: T::StableCoinAssetId::get(),
+				initial_amount: stable_asset_reserve,
+				initial_price: stable_asset_price,
+			});
+
+			Self::deposit_event(Event::TokenAdded {
+				asset_id: T::NativeAssetId::get(),
+				initial_amount: native_asset_reserve,
+				initial_price: native_asset_price,
+			});
+
+			Ok(())
+		}
+
+		/// Add new token to omnipool in quantity `amount` at price `initial_price`
+		///
+		/// Can be called only after pool is initialized, otherwise it returns `NoStableAssetInPool`
+		///
+		/// Initial liquidity must be transferred to pool's account for this new token manually prior to calling `add_token`.
+		///
+		/// Initial liquidity is pool's account balance of the token.
+		///
+		/// Position NFT token is minted for `position_owner`.
+		///
+		/// Parameters:
+		/// - `asset`: The identifier of the new asset added to the pool. Must be registered in Asset registry
+		/// - `initial_price`: Initial price
+		/// - `position_owner`: account id for which share are distributed in form on NFT
+		///
+		/// Emits `TokenAdded` event when successful.
+		///
+		#[pallet::weight(<T as Config>::WeightInfo::add_token())]
+		#[transactional]
+		pub fn add_token(
+			origin: OriginFor<T>,
+			asset: T::AssetId,
+			initial_price: Price,
+			position_owner: T::AccountId,
+		) -> DispatchResult {
+			T::AddTokenOrigin::ensure_origin(origin)?;
+
```

Comment:
added some comments to divide some functions into 3 parts - preconditions, main, post - state updates

Reply:
so you are finally ready to get rid of empty lines as well ? :)

Reply:
> so you are finally ready to get rid of the empty lines? :) the change just increased the size of the method even more

;-)  if we add it to our code guidelines, then i would consider it.

I might remove some where the statements do the same things ( 2 transfers, 2 storage reads/writes - usually i do this), otherwise, i prefer with empty lines.

Reply:
I agree that the implementation might look bit too long but the "meat of it" like you named it is just one or two lines - the rest is precheck and state updates. And it is the way it must be done, bit unfortunate that there are many things to check and update.